smalltalksrc/VMMaker/CogObjectRepresentationForSpur.class.st

```diff
@@ -2913,6 +2913,101 @@ CogObjectRepresentationForSpur >> genVarIndexCallStoreTrampoline [
 				cogit annotateBytecode: cogit Label ].
 ]
 
+{ #category : #'as yet unclassified' }
+CogObjectRepresentationForSpur >> gen_primitiveAtByteArray [
+	"AutoGenerated by Druid"
+
+	| jump3 jump1 currentBlock jump4 jump2 |
+	cogit TstCq: 7 R: ReceiverResultReg.
+	jump1 := cogit JumpNonZero: 0.
+	cogit MoveR: Arg0Reg R: TempReg.
+	cogit ArithmeticShiftRightCq: 3 R: TempReg.
+	cogit MoveM64: 0 r: ReceiverResultReg R: ClassReg.
+	cogit ArithmeticShiftRightCq: 24 R: ClassReg.
+	cogit AndCq: 31 R: ClassReg.
+	cogit CmpCq: 16 R: ClassReg.
+	jump2 := cogit JumpLess: 0.
+	cogit MoveMb: 7 r: ReceiverResultReg R: SendNumArgsReg.
+	cogit AndCq: 255 R: SendNumArgsReg.
+	cogit CmpCq: 255 R: SendNumArgsReg.
+	jump3 := cogit JumpZero: 0.
+	jump4 := cogit Jump: 0.
+	currentBlock := cogit Label.
+	jump3 jmpTarget: currentBlock.
+	cogit MoveM64: -8 r: ReceiverResultReg R: SendNumArgsReg.
+	cogit LogicalShiftLeftCq: 8 R: SendNumArgsReg.
+	cogit ArithmeticShiftRightCq: 8 R: SendNumArgsReg.
+	currentBlock := cogit Label.
+	jump4 jmpTarget: currentBlock.
+	cogit LogicalShiftLeftCq: 3 R: SendNumArgsReg.
+	cogit AndCq: 7 R: ClassReg.
+	cogit SubR: ClassReg R: SendNumArgsReg.
+	cogit CmpCq: 1 R: TempReg.
+	jump4 := cogit JumpLess: 0.
+	cogit CmpR: SendNumArgsReg R: TempReg.
+	jump3 := cogit JumpGreater: 0.
+	cogit MoveR: ReceiverResultReg R: SendNumArgsReg.
+	cogit AddR: SendNumArgsReg R: TempReg.
+	cogit AddCq:  -1 R: TempReg.
+	cogit MoveMb: 8 r: TempReg R: SendNumArgsReg.
+	cogit AndCq: 255 R: SendNumArgsReg.
+	cogit LogicalShiftLeftCq: 3 R: SendNumArgsReg.
+	cogit AddCq: 1 R: SendNumArgsReg.
+	cogit MoveR: SendNumArgsReg R: ReceiverResultReg.
+	cogit genPrimReturn.
+	currentBlock := cogit Label.
+	jump1 jmpTarget: currentBlock.
+	jump2 jmpTarget: currentBlock.
+	jump4 jmpTarget: currentBlock.
+	jump3 jmpTarget: currentBlock.
+	^ CompletePrimitive
+]
+
+{ #category : #'as yet unclassified' }
+CogObjectRepresentationForSpur >> gen_primitiveAtPointerArray [
+	"AutoGenerated by Druid"
+
+	| jump3 jump1 currentBlock jump4 jump2 |
+	cogit TstCq: 7 R: ReceiverResultReg.
+	jump1 := cogit JumpNonZero: 0.
+	cogit MoveR: Arg0Reg R: TempReg.
+	cogit ArithmeticShiftRightCq: 3 R: TempReg.
+	cogit MoveM64: 0 r: ReceiverResultReg R: ClassReg.
+	cogit ArithmeticShiftRightCq: 24 R: ClassReg.
+	cogit AndCq: 31 R: ClassReg.
+	cogit CmpCq: 2 R: ClassReg.
+	jump2 := cogit JumpNonZero: 0.
+	cogit MoveMb: 7 r: ReceiverResultReg R: ClassReg.
+	cogit AndCq: 255 R: ClassReg.
+	cogit CmpCq: 255 R: ClassReg.
+	jump3 := cogit JumpZero: 0.
+	jump4 := cogit Jump: 0.
+	currentBlock := cogit Label.
+	jump3 jmpTarget: currentBlock.
+	cogit MoveM64: -8 r: ReceiverResultReg R: ClassReg.
+	cogit LogicalShiftLeftCq: 8 R: ClassReg.
+	cogit ArithmeticShiftRightCq: 8 R: ClassReg.
+	currentBlock := cogit Label.
+	jump4 jmpTarget: currentBlock.
+	cogit CmpCq: 1 R: TempReg.
+	jump4 := cogit JumpLess: 0.
+	cogit CmpR: ClassReg R: TempReg.
+	jump3 := cogit JumpGreater: 0.
+	cogit SubCq: 1 R: TempReg.
```

Comment:
An interesting not here is that we know we cannot fail anymore. So here we should be able to use the Receiver register and override it as soon as it is not used anymore.
This 

```smalltalk
cogit SubCq: 1 R: TempReg.
	cogit LogicalShiftLeftCq: 3 R: TempReg.
	cogit MoveR: ReceiverResultReg R: ClassReg.
	cogit AddR: ClassReg R: TempReg.
	cogit MoveM64: 8 r: TempReg R: ClassReg.
	cogit MoveR: ClassReg R: ReceiverResultReg.
```

could eventually be compiled as

```smalltalk
        cogit SubCq: 1 R: TempReg.
	cogit LogicalShiftLeftCq: 3 R: TempReg.
	cogit AddR: ReceiverResultReg. R: TempReg.
	cogit MoveM64: 8 r: TempReg R: ReceiverResultReg.
```